### PR TITLE
208 logger spi

### PR DESCRIPTION
**Pull request by maor226**

### CODE REVIEW

File: src/main/java/il/ac/bgu/cs/bp/bpjs/execution/jsproxy/DefaultBpLog.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.io.PrintStream;
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.Set;
+
+/**
+ * Simple logging mechanism for {@link BProgram}s.
+ *
+ * @author michael
```
Comment (michbarsinai) [inline src/main/java/il/ac/bgu/cs/bp/bpjs/execution/jsproxy/DefaultBpLog.java:37]: I didn't author this :-)

File: src/main/java/il/ac/bgu/cs/bp/bpjs/execution/jsproxy/DefaultBpLog.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ */
+public class DefaultBpLog implements BpLog {
```
Comment (michbarsinai) [inline src/main/java/il/ac/bgu/cs/bp/bpjs/execution/jsproxy/DefaultBpLog.java:39]: I would change the name to "PrintStreamBpLog" or the like. The fact that it's a default logger is not a description of the class itself, but rather of BPjs.

File: src/main/java/il/ac/bgu/cs/bp/bpjs/model/BProgram.java [hunk 3]
```diff
@@ -504,7 +507,15 @@ public void setLogLevel( BpLog.LogLevel aLevel ) {
...
         }
     }
-    
+
+    /**
+     * Set the log implementation - logging called below the set level are logged using @BpLog implementation.
+     * @param logger The new logger implementation to use.
+     */
+    public void setLogger(BpLog logger){
+        preSetLogger = logger;
```
Comment (michbarsinai) [inline src/main/java/il/ac/bgu/cs/bp/bpjs/model/BProgram.java:516]: If I understand, once the BProgram is started, this set would have no effect. So in this case we also need to update the logger in `jsProxy`. (unless I'm missing something here).

File: README.md [hunk 4]
```diff
@@ -56,6 +56,7 @@ If you use BPjs in an academic work, please consider citing it as:
 * :arrow_up: Upgraded dependencies.
 * :arrow_up: `MapProxy`'s seed is now `protected`, so client code can use it directly ([#206](https://github.com/bThink-BGU/BPjs/issues/206)).
 * :sparkles: Dedicated exceptions for calling `bp.sync` and `bp.thread` outside of a bthread. ([#195](https://github.com/bThink-BGU/BPjs/issues/195), [#203](https://github.com/bThink-BGU/BPjs/issues/203))
+* :part_alternation_mark: `BpLog` is now an interface, and it's implementation can be changed by calling `BProgram.setLogger`. ([#208](https://github.com/bThink-BGU/BPjs/issues/208))
```
Comment (michbarsinai) [inline README.md:59]: Nice :-)

**Review by michbarsinai — CHANGES_REQUESTED** [review]
Very nice work! Some minor adjustments are needed. Also need to have a test that verifies that a new logger gets the data properly. 

For example, a test can create a bprogram that writes to the log in various levels, set a new logger (maybe one that aggregates messages in `List<String>`s), and then validate that the contents of the list in each logger level is as expected.

File: src/main/java/il/ac/bgu/cs/bp/bpjs/model/BProgram.java (hunk 3 above)
Comment (maor226) [reply src/main/java/il/ac/bgu/cs/bp/bpjs/model/BProgram.java]: we cannot change the log after initializing the b-program, I added an exception.

**Review by michbarsinai — APPROVED** [review]
:-)

**Pull request merged**